File: app/src/main/java/com/stedi/randomimagegenerator/app/other/CachedBus.java
package com.stedi.randomimagegenerator.app.other;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import com.squareup.otto.Bus;
import com.squareup.otto.DeadEvent;
import com.squareup.otto.ThreadEnforcer;
import com.stedi.randomimagegenerator.app.other.logger.Logger;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Set;

public class CachedBus extends Bus {
    private final LinkedList<Runnable> cache = new LinkedList<>();
    private final Set<Object> postDeadEvents = new HashSet<>();
    private final Thread creationThread = Thread.currentThread();

    private final Logger logger;

    private boolean locked;

    public CachedBus(@NonNull Logger logger) {
        this(ThreadEnforcer.MAIN, logger);
    }

    public CachedBus(@NonNull ThreadEnforcer enforcer, @NonNull Logger logger) {
        super(enforcer);
        this.logger = logger;
    }

    public void lock() {
        logger.log(this, "lock");
        ensureCreationThread();
        locked = true;
    }

    public void unlock() {
        logger.log(this, "unlock");
        ensureCreationThread();
        locked = false;
        releaseCache();
    }

    public void postDead(final Object event) {
        logger.log(this, "postDead");
        ensureCreationThread();
        postDeadEvents.add(event);
        post(event);
    }

    @Override
    public void post(final Object event) {
        logger.log(this, "posting " + event);
        ensureCreationThread();
        if (!locked && !(event instanceof DeadEvent)) {
            logger.log(this, "posting " + event + " successfully");
            super.post(event);
            postDeadEvents.remove(event);
        } else {
            logger.log(this, "posting " + event + " failed");
            Object actualEvent = event;
            if (event instanceof DeadEvent) {
                actualEvent = ((DeadEvent) event).event;
                if (postDeadEvents.contains(actualEvent)) {
                    logger.log(this, "ignoring postDead event " + actualEvent);
                    postDeadEvents.remove(actualEvent);
                    return;
                }
            }
            logger.log(this, "adding to the cache");
            Object actualEventF = actualEvent;
            cache.add(() -> post(actualEventF));
        }
    }

    private void releaseCache() {
        if (!cache.isEmpty()) {
            LinkedList<Runnable> release = new LinkedList<>(cache);
            cache.clear();
            while (!release.isEmpty())
                release.pollFirst().run();
        }
    }

    private void ensureCreationThread() {
        if (Thread.currentThread() != creationThread)
            throw new IllegalStateException("CachedBus requires the thread from which it was created");
    }

    @VisibleForTesting
    LinkedList<Runnable> getCache() {
        return cache;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/other/CachedBus.kt
package com.stedi.randomimagegenerator.app.other

import android.support.annotation.VisibleForTesting
import com.squareup.otto.Bus
import com.squareup.otto.DeadEvent
import com.squareup.otto.ThreadEnforcer
import timber.log.Timber
import java.util.*
import javax.inject.Singleton

@Singleton
class CachedBus(enforcer: ThreadEnforcer = ThreadEnforcer.MAIN) : Bus(enforcer) {

    private val cache = LinkedList<Runnable>()
    private val allowedDeadEvents = HashSet<Any>()
    private val creationThread = Thread.currentThread()

    private var locked: Boolean = false

    fun lock() {
        Timber.d("lock")
        ensureCreationThread()
        locked = true
    }

    fun unlock() {
        Timber.d("unlock")
        ensureCreationThread()
        locked = false
        releaseCache()
    }

    fun postDeadEvent(event: Any) {
        Timber.d("postDeadEvent")
        ensureCreationThread()
        allowedDeadEvents.add(event)
        post(event)
    }

    override fun post(event: Any) {
        Timber.d("posting $event")
        ensureCreationThread()
        if (!locked && event !is DeadEvent) {
            Timber.d("posting $event successfully")
            super.post(event)
            allowedDeadEvents.remove(event)
        } else {
            Timber.d("posting $event failed")
            var actualEvent = event
            if (event is DeadEvent) {
                actualEvent = event.event
                if (allowedDeadEvents.contains(actualEvent)) {
                    Timber.d("ignoring postDeadEvent event $actualEvent")
                    allowedDeadEvents.remove(actualEvent)
                    return
                }
            }
            Timber.d("adding to the cache")
            cache.add(Runnable { post(actualEvent) })
        }
    }

    private fun releaseCache() {
        if (!cache.isEmpty()) {
            val release = LinkedList(cache)
            cache.clear()
            while (!release.isEmpty()) {
                release.pollFirst().run()
            }
        }
    }

    private fun ensureCreationThread() {
        if (Thread.currentThread() !== creationThread) {
            throw IllegalStateException("CachedBus requires the thread from which it was created")
        }
    }

    @VisibleForTesting
    fun getCache() = cache

    @VisibleForTesting
    fun getAllowedDeadEvents() = allowedDeadEvents
}